Fix: enable Python 3.12 to parse templates

scottbarnes opened the pull request:
Closes #784.

This PR fixes the inability of `web.py` to parse templates on Python 3.12, and it does so by matching any unmatched `"` in a string.

In Python 3.12, changes to `tokenize.generate_tokens()` requires all quotes to be matched, and the `web.py` template parser hands unterminated strings to `tokenize.generate_tokens()` quite frequently.

This PR makes a tacit assumption that we should avoid rewriting the template parser if possible while also allowing `web.py` to parse templates on Python 3.12.

To that end, it handles every `TokenError` caused by unterminated string literals (i.e. unmatched `"` characters) by adding a `"` to the line where one is missing. Although this appears to work, I am aware that appending a `"` to each line that has an unmatched `"` is less than ideal. However, it seemed a very easy way to avoid rewriting the template parser. I am absolutely open to other approaches.

*IF* this approach looks as if it has legs, I can can look more into where that "extra" `"` goes, why it seems not to matter, and write some unit tests.

With regard to testing, I tested this with as many pages on Open Library that I could (via the local development environment), and it seems to work.

Edit: for the time being I added an exception to the upper limits on code complexity. If this solution looks worth pursuing, I can try to sort out the code complexity issue.

cclauss commented on web/template.py:
```diff
@@ -242,7 +242,7 @@ def read_keyword(self, text):
         return StatementNode(line.strip() + "\n"), text
 
-    def read_expr(self, text, escape=True):
+    def read_expr(self, text, escape=True):  # noqa: C901, PLR0915
```
Can you please add the string that was causing you trouble into a new test in tests/test_template.py so we can ensure that we do not have any regressions?

Please drop the noqas in favor of increasing the corresponding values at
https://github.com/webpy/webpy/blob/ed3e92cceb6ed870b224107ea653f48fa7fc2d0a/pyproject.toml#L86-L93

scottbarnes replied:
I did this, but I sure don't feel good about it! I also added commit to move some code out of `Parse` that would be easy to drop if undesired.

tfmorris commented on web/template.py:
```diff
@@ -306,23 +306,38 @@ def paren_expr():
+                    yield t
+
+            try:
+                yield from tokenize_text(text)
+            except tokenize.TokenError as e:
+                # Python 3.12+ compatibility: add a `"` suffix to terminate the string literal.
+                # See https://github.com/webpy/webpy/issues/784.
+                if "unterminated string literal" in str(e):
+                    fixed_text = text + '"'
+                    yield from tokenize_text(fixed_text)
```
Does `get_tokens()` raise an exception before returning any tokens? If not, it seems like you're going to get duplicate tokens with the retry strategy.

The current strategy seems like it's going to return a different string of tokens than before. It seems slightly cleaner to  use the same strategy that's currently used for returning white space (token type = -1) and create a token starting at the error position reported in the `TokenError` and running until the end of the string (or perhaps a token of type `token.ERRORTOKEN`).

I think that strategy could probably be used with any type of `TokenError` which is raised without having to check that it's an unterminated string literal.

Note that I haven't tested any of this!

scottbarnes replied:
I will try this. Thanks!

For what it is worth, at least at the `read_suite()` level, the existing `web.py` code on 3.11, and the PR code running on Python 3.12, produce the same outputs, or seem to, at the `read_suite()` level, though I have not dug into why that 'extra' `"` disappears.

I did take the output from some templates that were crashing on 3.12 and ran them both through 3.11 without the PR code, and 3.12 with the PR code, and ran `sort u` on them in `nvim`, and they were all identical.

For example, on Python 3.12 with this PR's code applied:
```python
Python 3.12.1 (main, Dec 19 2023, 03:09:45) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import web
>>> from web.template import Parser
>>> input_read_section = """$var title: Index
... <p><a href="$back">&larr; Back to Index</a></p>
... <ul>
... $for path, title in docs:
...     <li><a href="$path">$title</a></li>
... </ul>
... """
>>> Parser().read_suite(input_read_section)
[<var: title = join_('Index')>, <line: [t'<p><a href="', $back, t'">&larr; Back to Index</a></p>\n']>, <line: [t'<ul>\n']>, <block: 'for path, title in docs:', [<line: [t'<li><a href="', $path, t'">', $title, t'</a></li>\n']>]>, <line: [t'</ul>\n']>]
```

On Python 3.11, with the existing `web.py` code:
```python
Python 3.11.7 (main, Jan  5 2024, 22:06:16) [GCC 13.2.1 20230801]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.19.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import web

In [2]: from web.template import Parser

In [3]: input_read_section = """$var title: Index
   ...: <p><a href="$back">&larr; Back to Index</a></p>
   ...: <ul>
   ...: $for path, title in docs:
   ...:     <li><a href="$path">$title</a></li>
   ...: </ul>
   ...: """

In [4]: Parser().read_suite(input_read_section)
Out[4]: [<var: title = join_('Index')>, <line: [t'<p><a href="', $back, t'">&larr; Back to Index</a></p>\n']>, <line: [t'<ul>\n']>, <block: 'for path, title in docs:', [<line: [t'<li><a href="', $path, t'">', $title, t'</a></li>\n']>]>, <line: [t'</ul>\n']>]
```

I am not in love with adding the single `"`, though. I will look into the suggestion of creating a new token, and put some more time into understanding how these tokens are used, and how much is read (as not every part of a line is sent to this function -- it's the components, as created by `read_node()`, that start with `$` that are sent to `read_expr()`, and hence have `tokenize.generate_tokens()` called on them.

I would have to verify it, but my suspicion is that because the lines seem to end with new line characters, when I append a `"`, it's after a newline that may be getting stripped. But I could be completely wrong. I will both try the suggestion of creating a new token, and also understand why exactly the outputs are identical in 3.11 without the PR, and in 3.12 with the PR.

tfmorris commented:
Looks like that approach will work. https://github.com/webpy/webpy/commit/9b1403bf1b8b7fd0028a8fae9759061c783ff3f1

I don't think any of the extra tokens are actually used, so they probably could even just be skipped, but I didn't test that.

I did add a simple unit test for the failing case (and confirmed it works on 3.11 and fails on 3.12 w/o the patch).

scottbarnes commented:
You have the skills to pay the bills, @tfmorris. :D

Would you like to submit a new PR, or should I incorporate the changes you suggested, and tried out, into this one? I am not quite sure where https://github.com/webpy/webpy/commit/9b1403bf1b8b7fd0028a8fae9759061c783ff3f1 is living.

tfmorris commented:
It's your PR. I just tweaked it a bit. The commit is in my forked repo. Let me know if you have any trouble accessing it.

scottbarnes commented:
@tfmorris, @cclauss, I hope I implemented the suggestions correctly.

@tfmorris, as you suspected, the 'extra' tokens are not used, which probably explains why adding the extra `"` worked also. I think what you suggested is better though. :) I opted to leave the tokens in, even if not needed, as that mimics previous behavior and perhaps someone needs them somehow.

@cclauss, I added an optional commit to try to improve the code complexity situation a bit, but it may not make a realistic difference since it's just copy/pasting code to another place. The commit would be easy to drop.

I added the test @tfmorris made, and another one that may have pointless repetition, but my thinking was it just took a few seconds to simulate parsing of a template, at least for the `Parser.read_expr()` function, which previously worked on Python 3.11 and broke on Python 3.12.

mekarpeles commented:
Does this seem ready to merge? Thank you @scottbarnes for your work on getting web.py ready for Python 3.12!

tfmorris commented on web/template.py:
```diff
@@ -54,6 +56,41 @@
 from collections.abc import MutableMapping
 
 
+class BetterIter:
+    """Iterator like object with 2 support for 2 look aheads."""
+
+    def __init__(self, items):
+        self.iteritems = iter(items)
+        self.items = []
+        self.position = 0
+        self.current_item = None
+
+    def lookahead(self):
+        if len(self.items) <= self.position:
+            self.items.append(self._next())
+        return self.items[self.position]
+
+    def _next(self):
+        try:
+            return next(self.iteritems)
+        except StopIteration:
+            return None
+
+    def lookahead2(self):
+        if len(self.items) <= self.position + 1:
+            self.items.append(self._next())
+        return self.items[self.position + 1]
+
+    def __next__(self):
+        self.current_item = self.lookahead()
+        self.position += 1
+        return self.current_item
```
It doesn't look like `lookahead2()` is used at all, so this could be replaced by [`more-itertools.peekable`](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.peekable) but I don't feel strongly about it (and it does introduce another dependency)

tfmorris replied:
Oops, scratch that, lookahead2 actually does appear to be used and I missed it. Sorry!

tfmorris replied:
But `peekable` can still be used with `peekable[1]` replacing `lookahead2()`. See the commit referenced below.

tfmorris commented on pyproject.toml:
```diff
@@ -83,14 +83,14 @@ show-source = true
 max-branches = 17  # default is 12
 max-returns = 8  # default is 6
-max-statements = 51  # default is 50
+max-statements = 50  # default is 50
```
Can this just be dropped if it's the same as the default?

tfmorris reviewed (approved):
A couple of minor optional comments (although perhaps eliminating `BetterIterator` would help with your code shuffling, which seems a little arbitrary).

IM999MaxGood commented:
https://stackoverflow.com/questions/77997160/using-templates-in-web-py

IM999MaxGood commented:
💐hello
I'm sorry my English language is not good
I have problem with using web.py 
[my code in github.com](https://github.com/IM999MaxGood/Question_web-page-with-mysql-and-web2) my problems [my problems](https://stackoverflow.com/questions/77997160/using-templates-in-web-py)

I use changes in my code [changes](https://github.com/webpy/webpy/pull/785/commits/a23d1250d1f12fd3f6d0733bf49dc9649a3cd028)
errors in templates solved
but I have still error in using (where="id=$id") in my model.py even with python 3.11.2

scottbarnes commented:
> I use changes in my code [changes](https://github.com/webpy/webpy/pull/785/commits/a23d1250d1f12fd3f6d0733bf49dc9649a3cd028) errors in templates solved but I have still error in using (where="id=$id") in my model.py even with python 3.11.2

Are you getting the same error in Python 3.11.2? If not, it may be best to open a new issue.

That aside, it's possible this is a string interpolation issue that may be fixed by something like:
```diff
- db.delete("item", where="id=$id")
+ db.delete("item", where=f"id={id}")
```

tfmorris commented:
There's an off-by-one bug in my fix. I'm generating an update now.

> Message ID: ***@***.***>
>

scottbarnes commented:
Thank you for catching that, @tfmorris, and for contributing a fix. `peekable` is pretty slick. I saw that `more_itertools` is a dependency of `cheroot`, so I didn't update `requirements.txt`, but as I suppose `cheroot` may in theory drop the dependency, I will add it to `webpy` in a moment.

cclauss commented:
Thanks for the teamwork on this effort!
